### PR TITLE
Streamline and automate the versioning and publishing of romcal

### PR DESCRIPTION
**Pull request by pejulian**
As described in [my comments](https://github.com/romcal/romcal/issues/79#issuecomment-581043070) on #79, this PR aims to make life easier for contributors of `romcal` by handling:

- semantic versioning
- distribution tagging
- publishing to npm

Once enacted, we should be able to rely on the automated infrastructure to handle all the above for us.

The machine user saves contributors from:
- needed to have access to Travis and NPM for build and publishing
- having to manually keep track of semantic versioning numbers and formats
- having to set up their machines to publish new versions of the module
- suffer setbacks from a non responding human party with all the needed administrative rights to publish the module (i.e. me)

### CODE REVIEW

**Comment by emagnier** [comment]
Great! I'll try to do a review tonight or tomorrow on it.
Also you mention some remaining todos in this [comment](https://github.com/romcal/romcal/issues/79#issuecomment-581043070), do you plan to add them here or we can already merge this PR before theses future additions?

File: scripts/bump-version.sh
Comment (tukusejssirs) [inline scripts/bump-version.sh]:
```bash
    # If the tag version is higher than the "latest" version in npm,
    # then no new stable version has been published from master yet
    # and we should continue incrementing the "preid" based on the tag version 
```

File: scripts/semver.sh
Comment (tukusejssirs) [inline scripts/semver.sh]: I believe that `comparition` is misspelled `comparison`. Still, I think it might sound better when re-worded.

File: scripts/semver.sh
Comment (tukusejssirs) [inline scripts/semver.sh]: Why is `chmod`-ing required? Git repos keep the permissions, therefore _if you commit_ this file with `+x` (or even `a+x`), it will be executable when cloned and checked out on any machine.

File: scripts/semver.sh
Comment (tukusejssirs) [inline scripts/semver.sh]: As I see it, you have copied this script from somewhere. Still, I’d capitalise the first letters of all the comments.

And add the definite article (`the`) before ordinal number `first`.

File: scripts/semver.sh
Comment (tukusejssirs) [inline scripts/semver.sh]: I’d remove `Case` from the beginning. `When A is not a pre-release` is suffiecient (also note the indefinite article `a` before `pre-release`).

File: scripts/semver.sh
Comment (tukusejssirs) [inline scripts/semver.sh]: I’d go with `When pre-release A exists and is greater then pre-release B`.

File: scripts/semver.sh
Comment (tukusejssirs) [inline scripts/semver.sh]:
```
Extract `rc` numbers
````

File: scripts/semver.sh
Comment (tukusejssirs) [inline scripts/semver.sh]: `Retrun` → `Return`

**Review by tukusejssirs — CHANGES_REQUESTED** [review]

**Comment by pejulian** [comment]
> Great! I'll try to do a review tonight or tomorrow on it.
> Also you mention some remaining todos in this [comment](https://github.com/romcal/romcal/issues/79#issuecomment-581043070), do you plan to add them here or we can already merge this PR before theses future additions?

Hey @emagnier,

I just added an enhancement to the script where we can use the identifier `[USE_CUSTOM_SEMVER]` in a commit message to skip the auto incrementing mechanism to prefer whatever value the user supplied in the `version` field of `package.json`. As with all other human operations, human error is non avoidable and the build will fail if an incompatible version was supplied (i.e. lower version that what is already published). As such, it will be the user's responsibility to check the latest version published on the given tag(s) and supply a higher (and valid semantic) version.

I think what we have now is enough to put into general use.
The other "todo's" for git tagging and auto change log can be merged in at some later point when I have them ready.

File: scripts/bump-version.sh
Comment (pejulian) [reply scripts/bump-version.sh]: Hi @tukusejssirs, I've amended the comments here as per your recommendation.

File: scripts/semver.sh
Comment (pejulian) [reply scripts/semver.sh]: Hi @tukusejssirs,

For this and the comments below in respect to `semver.sh`, I think I'll leave the code and comments as-is because I did not write it. I could have but it's some pretty tedious stuff! So I found a code written by someone else that worked and decided to use that instead.

Further to this, I've also just updated the content of `semver.sh` with a more complete library that I found: 
https://github.com/fsaintjacques/semver-tool

File: scripts/semver.sh
Comment (pejulian) [reply scripts/semver.sh]: I'm nor sure why but I ran into some permission issues on TravisCI without explicitly adjusting file permissions.

Then I found this:
https://stackoverflow.com/questions/39666585/does-git-store-the-read-write-execute-permissions-for-files

It could be a safety precaution to adjust permissions when needed to avoid relying on Git entirely.

**Comment by pejulian** [comment]
You may have noticed that CI checks for this PR have been failing. That is so because I had previously set up a travis-ci.com (pro) account that triggered builds for this repository. The builds fail on the deploy step because the new `travis.yml` and corresponding bash scripts require environment variables that are now only available via the machine user travis-ci.org instance. This mechanism ensures that only 1 user is ever only responsible for build and publishing related activities for `romcal`. I have deactivated the old travis-ci.com instance on `romcal`. Future CI reports should now only come from the machine user travis instance.

**Review by emagnier — APPROVED** [review]
For me it's a LGTM. As soon as @tukusejssirs approve this PR, we can merge it.
I'm waiting to merge other pending PRs, I'll merge them when that one is merged. So we can test this new automated process.

**Comment by tukusejssirs** [comment]
> [@pejulian] As with all other human operations, human error is non avoidable and the build will fail if an incompatible version was supplied (i.e. lower version that what is already published). As such, it will be the user's responsibility to check the latest version published on the given tag(s) and supply a higher (and valid semantic) version.

I think that the custom version can (and should) be checked against the latest released version of the particular ‘release type’ (i.e. stable, beta, canary). Even more: I think that the user could just specify that they want to release the next major release and the `semver` should auto-calculate the next major version number (curretly that would be `2.0.0`). This could apply to minor and patch numbers, as well as beta/canary numbers.

**Review by tukusejssirs — APPROVED** [review]

**Comment by pejulian** [comment]
> > [@pejulian] As with all other human operations, human error is non avoidable and the build will fail if an incompatible version was supplied (i.e. lower version that what is already published). As such, it will be the user's responsibility to check the latest version published on the given tag(s) and supply a higher (and valid semantic) version.
> 
> I think that the custom version can (and should) be checked against the latest released version of the particular ‘release type’ (i.e. stable, beta, canary). Even more: I think that the user could just specify that they want to release the next major release and the `semver` should auto-calculate the next major version number (curretly that would be `2.0.0`). This could apply to minor and patch numbers, as well as beta/canary numbers.
> 
> **Update:** However, my suggestion above should not be taken as merge blocker, i.e. the script might be improved at a later stage (in that case, this should be moved into the `backlog`).

These are great suggestions @tukusejssirs 

It should be possible to expand the regex that matches the `[USE_CUSTOM_SEMVER]` substring to include a custom `semver` value itself (i,e. `[USE_CUSTOM_SEMVER:major|minor|patch|...]`) which will then be checked to ensure that it is a valid `semver` value before being applied to the package.

I'll look into this when I get some time.

**Pull request merged**
